### DISCUSSION TITLE
Creating on the fly json objects

### DISCUSSION THREAD

**Discussion by GJanos** [stephenberry/glaze#302] [General]
In my current assignment, I would like to construct JSON objects on the fly for logging purposes. It would be necessary so that I can log anything and everything without much effort or just for debugging purposes. 
Performance is really important in my case so I choose glaze as my ser/deserialization/json creation library. 
When creating a json object like this:
```c++
glz::json_t json = {
    {"pi", 3.141},
    {"happy", true},
    {"name", "Stephen"},
    {"nothing", nullptr},
    {"answer", {{"everything", 42.0}}},
    {"list", {1.0, 0.0, 2.0}},
    {"object", {
        {"currency", "USD"},
        {"value", 42.99}
        }
    }
};
```
It is said in the description of the _"If absolutely nothing is known about the JSON structure, then [glz::json_t] may be helpful, but it **comes at a hefty performance cost due to the heavy use of dynamic memory allocations.**"_ Indeed, after testing it out, this approached seemed slower than a purely stack memory allocated libraries solution (boost::json::value). I know it is not fair to compare these 2 options, but that is why I wrote this.

**Is there a way for me to construct glaze::json objects faster and with less memory allocations?** Could the smaller object be purely stack allocated and for bigger objects maybe pre-allocated heap could be a solution. If a solution already exists, please give me a short example.

Thank you for your time!

**Comment by stephenberry** [comment]
Thanks for bringing up your specific problem. The best solution, as you say, would be to stack allocate these objects. The `glz::json_t` uses a std::unordered_map, which is more expensive for your logging purpose.

What we probably want to support is the ability to use `glz::object` tuples directly. For example: 
```c++
auto obj = glz::object("pi", 3.141, "happy", true, "name", "Stephen");
std::string s{};
glz::write_json(obj, s);
```

We aren't far from supporting this, but I'm not sure when I'll be able to add this feature. However, your request has put this on my radar and I do think it is worthwhile. It would make logging extremely efficient without needing to write glz::meta structures.

Thanks for your feedback, sorry we don't have an immediate solution. But, I will turn this into an issue to address.

**Reply by GJanos** [reply]
Thank you for your fast and kind response, and for the detailed explanation! I am happy to hear that this question got into your radar! Hope it gets added soon! Wish you the best of luck!

**Reply by stephenberry** [reply]
Feature has been added, see #303

**Comment by GJanos** [comment]
Dear stephenberry! 
First of all, thank you again for reacting to my discussion so fast and adding a feature to support it. I have tried to work with it and have a few points, that I would like to ask and clarify.

If I created an object like this:
```c++
glz::obj{"pi", 3.141, "happy", true, "name", "Stephen", "arr", glz::arr{"Hello", "World", 2}});
```
I believe there would be no dynamic memory allocations (I am not sure, only tested the performance, but like this it was significantly faster).

Although on the other hand, if I created the object like this:
```c++
std::vector<int> vec = {1, 2, 3};
std::map<std::string, int> map = {{"a", 1},{"b", 2},{"c", 3}};
glz::obj{"pi", 3.141, "happy", true, "name", "Stephen", "arr", glz::arr{"Hello", "World", 2},
                               "map", map,
                               "vector", vec};
```
The performance plummeted. ( but it might just be due to the object being larger itself...)
Do you support the glz::map type in glz as well, or just the glz::array? I was not able to find it, but it would be a great addition, if not already added yet. ect: 
```c++
glz::obj{"pi", 3.141, "happy", true, "name", "Stephen", "arr", glz::arr{"Hello", "World", 2},
                               "map", glz::map{{"a", 1},{"b", 2},{"c", 3}}};
```

Other difficulty arise when I tried to pass the glz::obj as a function parameter. Because of its variadic template definition it gets really hard to pass that object around between functions.
```c++
template <typename... Args>
  static std::string serialize_data(glz::obj<Args...> const &data)
  {
      return glz::write_json(data);
  }
```
Could you somehow change this type or maybe put a wrapper class/struct around it that might make it easier to use?

Thank you for your time, looking forward to your answer!

**Reply by stephenberry** [reply]
@jSo, you need to be wary of using anything that would heap allocate in this context. Or, use `static thread_local` storage so that you don't keep allocating every time you build these structures.

For your example you could just write:
```c++
glz::arr vec = {1, 2, 3};
glz::obj map = {"a", 1,"b", 2,"c", 3};
auto obj = glz::obj{"pi", 3.141, "happy", true, "name", "Stephen", "map", map, "arr", glz::arr{"Hello", "World", 2}, "vec", vec};
```

Now everything is stack allocated and you get your performance back.

You can still use `std::vector` and `std::map` in this context, but you'l want to make sure they don't keep getting reallocated. So, you can store them in a class that you keep alive, or you can use `static thread_local` if you just want them locally in a function.

```c++
static thread_local std::vector<int> vec;
vec = {1, 2, 3};
static thread_local std::map<std::string, int> map;
map["a"] = 1;
map["b"] = 2;
map["c"] = 3;
auto obj = glz::obj{"pi", 3.141, "happy", true, "name", "Stephen", "map", map, "arr", glz::arr{"Hello", "World", 2}, "vec", vec};
```

Now the `vector` and `map` memory will be reused every time you call this code and you'll get much better performance (almost as fast as the first version).

As to your issue of passing around the object, because it is a tuple that depends on the input types I would recommend some type erasure if you can't stay within a templated context.

How you pass around the object depends on how you intend to use the object. If you just intend to serialize the object, you can pass around a `std::function<void(std::string&)>` that will type erase the `glz::obj` and invoke serialization whenever you like. You can also store this `std::function` in a `std::shared_ptr` if you need to share it to avoid copies. There are a ton of type erasure techniques you could use here, from using a `std::any` with an any-visitor pattern, to making a virtual interface to type erased storage. If you can explain further your use case of passing around the object, I could point you in a more specific direction.

Also, I would recommend reusing your buffer memory in your case to improve performance. Your example above calls `return glz::write_json(data);`, but this has to heap allocate a std::string with every call.

Instead, you can reuse the same std::string by storing it somewhere or using `static thread_local` again:
```c++
static thread_local std::string buffer;
glz::write_json(data, buffer);
return buffer;
```

**Reply by GJanos** [reply]
Dear @stephenberry !
Thank you for your response, I will use and take into account your advice!

**Reply by GJanos** [reply]
Dear @stephenberry !
This is unrelated to my previous post, I just did not want to spam your discussion page. 
Does glaze have a functionality where it can quickly and efficiently add 2 glz::json or glz::obj types? Or is it possible to concatenate 2 strings that are already in json format?

Thank you for your response!

**Reply by stephenberry** [reply]
@GJanos, So, you're wanting to merge two objects into the same object? Or are they separate objects, just part of the same JSON structure?

**Reply by GJanos** [reply]
I have a class named Log_info_data, that stores the loggable informations metadata. I want to attach this as an extra to whatewer I have to log by default (serializaed object, serialized glz::obj)...
And my goal would be to make a new json object, that could be serialized from the Log_info_data object and the object that I have to log.

New json object = Log_info_data(...) + Any_type_of_obj_to_be_logged();
std::string result = object.serialize();

**Reply by GJanos** [reply]
Dear @stephenberry you have not answered my previous post, could you help me solve my issue, or just tell me if it is not implemented yet.

**Reply by stephenberry** [reply]
Hi @GJanos, 
Merging two objects is not currently supported, there is actually an open issue related to this.
I definitely think this is a necessary feature, especially for logging, I'm just focused on some other updates to Glaze right now.
I'm trying to think through the best way to handle this as well.
I'm thinking the syntax will look like:
```c++
glz::merge{ glz::obj{...}, glz::obj{...} };

```

This include lvalue objects as well to solve your problem. Or we could overload the `+` operator, which I like the idea of.

But, we're about to add ranges (std::views) support that merges `std::pair` values. I'm thinking this syntax should perhaps also merge glz::obj values. Though I'm a bit torn because we may want ranges that produce a JSON array of objects.

I'm thinking it's better to overload the `+` operator and provide a `glz::merge` type.

Will be thinking about this more, especially in regards to our current pull request. And, I'll add to this thread when the feature has been completed. Thanks for bringing this to my attention.

**Reply by GJanos** [reply]
Thank you for your response, I will be looking forward to your implementation!

**Reply by stephenberry** [reply]
@GJanos, I just merged (#333) in the addition of `glz::merge`. It probably should include additional logic for merging other types, like arrays, but see if this solves your immediate problem.

Here's the example unit test:
```c++
glz::obj obj0{"pi", 3.141};
glz::obj obj1{"happy", true};
glz::obj obj2{"arr", glz::arr{"Hello", "World", 2}};
auto merged = glz::merge{ obj0, obj1, obj2 };
std::string s{};
glz::write_json(merged, s);

expect(s == R"({"pi":3.141,"happy":true,"arr":["Hello","World",2]})");
```

**Reply by stephenberry** [reply]
I merged a follow on change that forces it to take references to avoid copies when constructing `glz::merge`, however this prevents rvalue references to be passed to `merge`, which needs to be fixed in the future.

**Reply by GJanos** [reply]
Dear @stephenberry !
If I am correct glaze currently does not have  way to create glz::obj from user defined objects. If this were possible than logging would be super easy with glaze:

```c++

bool Elastic_handler::insert_trade_data(Log_data_info &&log_info, Trade_data &&trade_data) {

    glz::obj obj0 = log_info;
    glz::obj obj1 = trade_data;
    auto merged = glz::merge{ obj0, obj1};
    std::string s{};
    glz::write_json(merged, s);

    return send_curl_request(s.c_str());
}

```
I know I might be asking for too much, but would something like this be achievable?
Thank you for your time!

**Reply by stephenberry** [reply]
@GJanos 
I can add this behavior, and will try to do so soon.

**Reply by GJanos** [reply]
@stephenberry 
Thank you! Let me know when that feature is ready ! :)

**Reply by stephenberry** [reply]
@GJanos,
The latest merges (#341 & #342) should solve your problems. `glz::merge` now works with other JSON object-like types. Also, `glz::merge` can now take rvalue references for cleaner syntax. And, `glz::arr` and `glz::obj` also avoid copies when passed lvalues, so performance should be better in some cases.

**Reply by GJanos** [reply]
Dear @stephenberry !

I made a small demo code that explains one of my problems. When I run this, instead of incrementing numbers getting printed, I only get 10 10 times.

```c++
#include <vector>
#include <glaze/glaze.hpp>
#include <fmt/core.h>

int main() {
    int cnt = 0;
    glz::obj o{"count", cnt};
    std::vector<decltype(o)> vec;
    for (cnt = 0; cnt < 10; ++cnt) {
        vec.emplace_back(glz::obj{"cnt", cnt});
    }
    for( const auto& i: vec ) {
        fmt::print("count value:{}\n", glz::tuplet::get<1>(i.value));
    }
}
```
Log output: 10 pieces of 10. No incrementing numbers, because of the glz::obj taking the cnt value as a const int&. 

Glz::obj taking references indeed improve performance, but sometimes occasional deep-copies are needed. Especially when a glz::obj would want to be logged, so it would get copied and the copy would be put to a backend threads spnc queue for later work. But because of a shallow-copy getting made, the values of the object inside of the queue would get corrupted.

For this reason it would be an awesome feature to be able to customize the glz::obj and maybe even your own types (but currently glz::obj is the priority) whether you want them shallow or deep-copied.

Glaze already supports something like:
```c++
glz::write<glz::opts{.prettify = true}>(obj, buffer);
So, applying this to glz::obj would be something like:
glz::obj<glz::opts{.deep_copy = true}>
```
Extra boolean template, or some other solution. And default could be the shallow copy option. This could be a object lvl setting or maybe an attribute lvl option.
```c++
template<>
struct glz::meta<Trade_data> {
    using T = Trade_data;
    static constexpr auto value = glz::object(
            "id", &T::id, "shallow",
            "name", &T::name, "deep",
            "price", &T::price, "shallow",
            "quantity", &T::quantity, "shallow",
            "exchange_name", &T::exchange_name, "deep") ;
};
```
And based on the template parameter call deep or shallow copy constructor on the attributes of the class.

Basically I would need a way to deep copy glz::obj on the stack. Not with moving the data or references.

**Reply by stephenberry** [reply]
@GJanos 

I'll address your simple example. If you specify copies when you constructor your `glz::obj`, then you can solve this demo problem as follows:
```c++
size_t cnt = 0;
glz::obj o{"count", size_t{cnt}};
std::vector<std::decay_t<decltype(o)>> vec;
for (; cnt < 10; ++cnt) {
   vec.emplace_back(glz::obj{"count", size_t{cnt}});
}
for (size_t i = 0; i < vec.size(); ++i) {
   expect(i == glz::tuplet::get<1>(vec[i].value));
}
```

This all works as expected because the counts are all copies as you want them. This may also work for your larger problem. If it doesn't, let me know and we can discuss the particular problem.
With `glz::obj` taking references I like the fact that copies must be explicitly defined by the user, so it is more obvious where there can be performance bottlenecks.

> Also, note my added `std::decay_t` to the object type stored in the vector. This is needed to avoid taking references.

**Comment by GJanos** [comment]
Dear @stephenberry !

I have found a possible bug. When I have a std::string containing a json, and I would like to convert that to a glz supported object, so later on I could merge that object with an another. I did not find any easy way to achieve this. 
I wanted something like:
```c++
glz::obj temp = glz::parse_json_str(string_json);

//found this approach
glz::json_t j;
glz::read_json(j,string_json);
```
This worked, however, when I tried to std::merge glz::obj and glz::json_t object (where the glz::json_t contained a MAP) parsing problems happened. This could be due to glz::obj not handling map types. But would be nice it these 2 types could be merged together easily OR make a glz::parse_json_str like function that immediately makes a glz::obj, and handles map types inside the string as well.

Made a small demo that illustrates the problem:
```c++
int main() {
    glz::obj o {"not", "important"};
    glz::obj o2 {"map", glz::obj{"a", 1, "b", 2, "c", 3}};
    std::string s = "{\"key1\":42,\"key2\":\"hello world\",\"map\":\"key\",\"v\":[1,2,3],\"m\":{\"a\":1,\"b\":2,\"c\":3}}";
    glz::json_t j;
    glz::read_json(j,s);
    auto test1 = glz::write_json(j);
    auto merged = glz::merge{o, j};
    auto test2 = glz::write_json(merged);
    auto merged2 = glz::merge{o, o2};
    auto test3 = glz::write_json(merged2);
    fmt::print("{}\n", test1);
    fmt::print("{}\n", test2);
    fmt::print("{}\n", test3);
//    {"key1":42,"key2":"hello world","m":{"a":1,"b":2,"c":3},"map":"key","v":[1,2,3]}
//    {"not":"important","key1":42,"key2":"hello world","m":"a":1,"b":2,"c":3,"map":"key","v":[1,2,3]}
//    {"not":"important","map":{"a":1,"b":2,"c":3}
}
```
Thank you for reading my message, I am looking forward to your reply for this and my previous comment.

**Reply by stephenberry** [reply]
@GJanos 
Thanks for pointing this out. There was a problem with nested objects within `glz::obj`. This has now been fixed (#349). I think this should solve your problem, but I do want to add reading directly into `glz::obj` in the future.
